File: apps/mac/Sources/ClipPlusMac/Sync/RemoteClipboardReceiveGuard.swift
import Foundation

final class RemoteClipboardReceiveGuard {
    private let lock = NSLock()
    private let imageFileSuppressionInterval: TimeInterval
    private var recentImageTimesByDeviceId: [String: Date] = [:]

    init(imageFileSuppressionInterval: TimeInterval = 15) {
        self.imageFileSuppressionInterval = imageFileSuppressionInterval
    }

    func recordRemoteImage(senderDeviceId: String, now: Date = Date()) {
        lock.lock()
        defer { lock.unlock() }

        recentImageTimesByDeviceId[senderDeviceId] = now
    }

    func shouldSuppressFileOfferAfterRecentImage(
        senderDeviceId: String,
        files: [FileTransferItem],
        now: Date = Date()
    ) -> Bool {
        guard isSingleImageFileOffer(files) else {
            return false
        }

        lock.lock()
        defer { lock.unlock() }

        guard let imageTime = recentImageTimesByDeviceId[senderDeviceId] else {
            return false
        }

        return now.timeIntervalSince(imageTime) <= imageFileSuppressionInterval
    }

    private func isSingleImageFileOffer(_ files: [FileTransferItem]) -> Bool {
        guard files.count == 1,
              let file = files.first,
              !file.isDirectory,
              file.byteSize > 0 else {
            return false
        }

        let fileExtension = URL(fileURLWithPath: file.relativePath).pathExtension.lowercased()
        return Self.imageFileExtensions.contains(fileExtension)
    }

    private static let imageFileExtensions: Set<String> = [
        "png",
        "jpg",
        "jpeg",
        "gif",
        "bmp",
        "tif",
        "tiff",
        "webp",
        "heic",
        "heif"
    ]
}

File: apps/windows/ClipPlus.Windows/Sync/RemoteClipboardReceiveGuard.cs
using System.IO;

namespace ClipPlus.Windows.Sync;

public sealed class RemoteClipboardReceiveGuard
{
    private static readonly HashSet<string> ImageFileExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".png",
        ".jpg",
        ".jpeg",
        ".gif",
        ".bmp",
        ".tif",
        ".tiff",
        ".webp",
        ".heic",
        ".heif"
    };

    private readonly object syncLock = new();
    private readonly TimeSpan imageFileSuppressionInterval;
    private readonly Dictionary<string, DateTimeOffset> recentImageTimesByDeviceId = new(StringComparer.Ordinal);

    public RemoteClipboardReceiveGuard()
        : this(TimeSpan.FromSeconds(15))
    {
    }

    public RemoteClipboardReceiveGuard(TimeSpan imageFileSuppressionInterval)
    {
        this.imageFileSuppressionInterval = imageFileSuppressionInterval;
    }

    public void RecordRemoteImage(string senderDeviceId)
    {
        RecordRemoteImage(senderDeviceId, DateTimeOffset.UtcNow);
    }

    public void RecordRemoteImage(string senderDeviceId, DateTimeOffset now)
    {
        lock (syncLock)
        {
            recentImageTimesByDeviceId[senderDeviceId] = now;
        }
    }

    public bool ShouldSuppressFileOfferAfterRecentImage(
        string senderDeviceId,
        IReadOnlyList<FileTransferItem> files)
    {
        return ShouldSuppressFileOfferAfterRecentImage(senderDeviceId, files, DateTimeOffset.UtcNow);
    }

    public bool ShouldSuppressFileOfferAfterRecentImage(
        string senderDeviceId,
        IReadOnlyList<FileTransferItem> files,
        DateTimeOffset now)
    {
        if (!IsSingleImageFileOffer(files))
        {
            return false;
        }

        lock (syncLock)
        {
            return recentImageTimesByDeviceId.TryGetValue(senderDeviceId, out var imageTime)
                && now - imageTime <= imageFileSuppressionInterval;
        }
    }

    private static bool IsSingleImageFileOffer(IReadOnlyList<FileTransferItem> files)
    {
        if (files.Count != 1)
        {
            return false;
        }

        var file = files[0];
        return !file.IsDirectory
            && file.ByteSize > 0
            && ImageFileExtensions.Contains(Path.GetExtension(file.RelativePath));
    }
}
